pkg/runner/test_runner.go

```diff
@@ -319,25 +313,38 @@ func (p *pipeline) route(terminateRoute <-chan struct{}, bundle test.TestStepBun
 			}
 		} // end of select statement
 
-		if err != nil {
-			break
+		if pendingTarget == nil {
+			if targets.Len() > 0 {
+				// More targets available to forward to the test step
+				pendingTarget = targets.Back().Value.(*target.Target)
+				ingressTarget[pendingTarget] = time.Now()
+				targets.Remove(targets.Back())
+				injectionWg.Add(1)
+				log.Debugf("writing target %v into test step", pendingTarget)
+				go targetWriter.writeTargetWithResult(terminate, pendingTarget, injectionChannels, &injectionWg)
```

Comment:
Since this is a pipelined design, the requirement was that the targets would move asynchronously and there wouldn't be any sync point where all of them need to be available before starting a plugin run. Whether all the targets should be collected, or the test can start on a per-target basis, or any other aggregation should happen, was left to decide to the plugin. Hence the targets flowing through channels.